drivers/node/vsphere/vsphere.go

```
@@ -0,0 +1,248 @@
+	logrus.Infof("Logging in to ESXi using: %s", login)
+
+	u, err := url.Parse(login)
+	if err != nil {
+		return fmt.Errorf("error parsing url %s", login)
+	}
+
+	v.ctx, v.cancel = context.WithCancel(context.Background())
+	//defer cancel()
+
+	c, err := govmomi.NewClient(v.ctx, u, true)
+	if err != nil {
+		return fmt.Errorf("logging in error: %s", err.Error())
+	}
+	logrus.Infof("Log in successful to vsphere:  %s:\n", v.vsphereHostIP)
+
+	f := find.NewFinder(c.Client, true)
+
+	// Find one and only datacenter
+	dc, err := f.DefaultDatacenter(v.ctx)
+	if err != nil {
+		return fmt.Errorf("Failed to find data center: %v", err)
+	}
+
+	// Make future calls local to this datacenter
+	f.SetDatacenter(dc)
+
+	// Find virtual machines in datacenter
+	vms, err := f.VirtualMachineList(v.ctx, "*")
+	if err != nil {
+		return fmt.Errorf("Failed to find any virtual machines on %s: %v", v.vsphereHostIP, err)
+	}
+
+	nodes := node.GetNodes()
+	for _, vm := range vms {
+		for _, n := range nodes {
+			if vm.Name() == n.Name {
+				if _, ok := vmMap[vm.Name()]; !ok {
+					vmMap[vm.Name()] = vm
+				}
+			}
+		}
+	}
+	return nil
+}
+
+// RebootVM reboots vsphere VM
+func (v *vsphere) RebootNode(n node.Node, options node.RebootNodeOpts) error {
+	if _, ok := vmMap[n.Name]; !ok {
+		return fmt.Errorf("Could not fetch VM for node: %s", n.Name)
+	}
+
+	vm := vmMap[n.Name]
+
+	logrus.Infof("Rebooting VM: %s  ", vm.Name())
+	err := vm.RebootGuest(v.ctx)
+	if err != nil {
+		return &node.ErrFailedToRebootNode{
+			Node:  n,
+			Cause: fmt.Sprintf("failed to reboot VM %s. cause %v", vm.Name(), err),
+		}
+	}
+	return nil
+}
+
+// PowerOnVM powers on the VM if not already on
+func (v *vsphere) PowerOnVM(n node.Node) error {
+	var err error
+	vm := vmMap[n.Name]
+
+	logrus.Infof("Powering on VM: %s  ", vm.Name())
+	tsk, err := vm.PowerOn(v.ctx)
+	if err != nil {
+		logrus.Errorf("Failed to power off %s: %v", vm.Name(), err)
+	} else {
```

redundant else here

---

yes, I need to return an error 
```if err != nil```